Heran IP

Heran9 opened the pull request:
### **Moss ChatBot**

> This is where everything starts😀

Moss can help you remember things. It's

- text-based
- simple
- easy to use

It contains commands such as:

1. todo, deadline, event (to add tasks to your task list) ``
2. list (check your task list) 
example of the implementaion of list :
`if (message.equals("list")) {
            // List all tasks
            System.out.println("____________________________________________________________");
            System.out.println("Here are the tasks in your list:");
            for (int i = 0; i < things.size(); i++) {
                System.out.println(i + 1 + "." + things.get(i).toString("x"));
            }
            System.out.println("____________________________________________________________");`
3. mark (mark things) you have already done
...

Here is the basic implementation of Moss:
```
public class Moss {
    /**
     * The list of tasks managed by the application.
     */
    static ArrayList<Task> things = new ArrayList<>();

    /**
     * The main method that initializes the application and handles user input.
     *
     * @param args Command-line arguments (not used in this application).
     * @throws MossException If there's an issue with loading tasks.
     */
    public static void main(String[] args) throws MossException {
        Storage storage = new Storage();

        try {
            things = (ArrayList<Task>) storage.loadTasks();
        } catch (Exception e) {
            throw new MossException("Failed to load tasks");
        }

        UI ui = new UI();
        ui.greeting();

        Scanner sc = new Scanner(System.in);
        String message = sc.nextLine();

        // Process user input until "bye" is entered
        while (!message.equals("bye")) {
            Parser parser = new Parser(message, things, storage);
            parser.execute(message, things, storage);
            message = sc.nextLine();
        }
        // Farewell message
        ui.bye();
    }

}
```
It is still a base model, and needs further development:

- [ X ] Base implementation
- [ ] Further improvement (still thinking about it)

Can reference [here](https://github.com/Heran9/ip/tree/master/src/main/java/Moss) to see the code implementation.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Heran9 committed:
add Moss.java, remove Duke.java

Heran9 committed:
add echo, make it a parrot

Heran9 committed:
add  Add, List make it a little notebook

Heran9 committed:
add Task class, change the message type String to Task, make it can be marked or unmarked.

Heran9 committed:
add different tasks

Heran9 committed:
automate test

Heran9 committed:
add exception handler

Heran9 committed:
add command delete

Heran9 committed:
add more javadocs

Heran9 committed:
removed emoji, there are some errors happened when using emoji in the string sentence at the automate test stage, dk why this happens

Heran9 committed:
read and write file

Heran9 committed:
Merge branch 'branch-Level-7'

Heran9 committed:
add datetime

Heran9 committed:
Merge branch 'branch-Level-8'

Heran9 committed:
make it more oop

Heran9 committed:
put all in a package

Heran9 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Heran9 committed:
Junit test

Heran9 committed:
create a jar file

Heran9 committed:
add javadocs

Heran9 committed:
fit Java coding standard

Heran9 committed:
add find tasks

Heran9 committed:
Merge branch 'branch-A-JavaDoc'

Heran9 committed:
deal with the conflicts

Heran9 committed:
Merge branch 'branch-Level-9'

Heran9 committed:
fix small bugs

jx124 commented on src/main/java/Moss/Task.java:
```diff
@@ -0,0 +1,85 @@
+     */
+    public void markDone() {
+        this.isDone = true;
+        System.out.println("____________________________________________________________");
+        System.out.println("Nice! I've marked this task as done:");
+        System.out.println(this.toString());
+        System.out.println("____________________________________________________________");
```
Perhaps the printing can be handled by the UI class?

jx124 commented on src/main/java/Moss/TaskList.java:
```diff
@@ -0,0 +1,151 @@
+     * @param storage The storage object for saving tasks.
+     * @throws MossException If there's an issue with processing commands or managing tasks.
+     */
+    public static void command(String message, ArrayList<Task> things, Storage storage) throws MossException {
```
Could use a verb for the name of this function since `command` by itself might be ambiguous. Perhaps `executeCommand`?

jx124 commented on src/main/java/Moss/Task.java:
```diff
@@ -0,0 +1,85 @@
+        return getStatusIcon() + " | " + this.description;
+    }
+
+    public String toString(String x) {
+        return getStatusIcon() + " | " + this.description;
+    }
```
Perhaps this could be named something different like `toListString` rather than overloading the `toString` function since it can be confusing as it differs in implementation only in `Event` and `Deadline`.

jx124 commented on src/test/java/moss/StorageTest.java:
```diff
@@ -0,0 +1,58 @@
+        List<Task> tasks = new ArrayList<>();
+        tasks.add(new ToDo("Buy groceries"));
+        tasks.add(new Deadline("Finish homework", LocalDate.parse("2023-09-15")));
+        tasks.add(new Event("Party", LocalDate.parse("2023-09-20"), LocalDate.parse("2023-09-22")));
```
I like this event!

jx124 commented on src/main/java/Moss/UI.java:
```diff
@@ -0,0 +1,37 @@
+    /**
+     * Displays a greeting message when the application starts.
+     */
+    public void greeting() {
```
Should start with a verb as well. Perhaps `showGreeting`?

jx124 commented on src/test/java/moss/MossTest.java:
```diff
@@ -0,0 +1,17 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class MossTest {
```
Not really sure if these tests are needed?

jx124 reviewed (commented):
Great attempt!

Heran9 committed:
Implement GUI

Heran9 committed:
Merge branch 'branch-Level-10'

Heran9 committed:
Add assertions to document and validate assumptions in Storage class

In this commit, we have added `assert` statements at various key points
in the `Storage` class to document and validate important assumptions.
These assertions enhance the robustness and maintainability of the code
by explicitly checking conditions that should be true during execution.

The assertions provide clear feedback in case any assumptions are
violated, making it easier to identify and fix issues during development
and maintenance.

Heran9 committed:
Remove some redundant code and delete some unused import statement.

Make some attributes final to make it safier.

Change a lot including remove redudant attribute, modify the constrcuter
since it is actually unnecessary.

Heran9 committed:
Merge pull request #1 from Heran9/branch-A-Assertions

Add assertions to document and validate assumptions in Storage class

Heran9 committed:
Merge branch 'master' into branch-A-CodeQuality

Heran9 committed:
Merge pull request #2 from Heran9/branch-A-CodeQuality

Branch a code quality

Heran9 committed:
Fix bugs by incorrect merges on github

Heran9 committed:
add extension C-help

Heran9 committed:
Remove unneceesary code, correct the format

Heran9 committed:
Add user guide and Ui.png

Heran9 committed:
Modify Readme.me and Ui.png

Heran9 committed:
Add more error handling

Heran9 committed:
correct the format

Heran9 committed:
Divide the long method into multiple short methods

Heran9 committed:
Modfiy Junittest and Userguide

Heran9 committed:
fix bug

Heran9 committed:
modify Ui
